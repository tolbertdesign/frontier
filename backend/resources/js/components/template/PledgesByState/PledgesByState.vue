<template>
  <div>
    <div class="mb-1 py-8 rounded-lg ">
      <UsaMap
        :active-states="selectedPledgedPlaces.states"
        class="text-grey-light"
      />
    </div>

    <div class="flex flex-col items-center">
      <span class="text-2xl">
        <span class="font-light">States:</span>
        <span class="font-bold">{{ statesCount }}</span>
      </span>
      <div class="relative">
        <select
          v-model="selectedPledgedPlaces"
          class="block appearance-none w-full py-1 px-4 pr-8 border border-primary bg-white rounded-md leading-tight focus:outline-none focus:bg-white focus:border-green-500"
          @change="trackSelect"
        >
          <option
            :value="myPledgedPlaces"
            data-label="My States"
          >
            My States
          </option>
          <option
            v-for="(classroom, index) in classrooms"
            :key="index"
            :value="classroom.pledgedPlaces"
            data-label="Specific Classroom"
          >
            {{ statesByClassroomLabel(classroom) }}
          </option>
          <option
            :value="program.pledgedPlaces"
            data-label="School States"
          >
            School States
          </option>
        </select>
        <div
          class="pointer-events-none absolute flex items-center px-2 pt-2"
          style="right: 0; top: 0"
        >
          <svg
            class="fill-current h-4 w-4"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 20 20"
          ><path d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z" /></svg>
        </div>
      </div>

      <p class="text-sm">
        Countries: <b>{{ [...selectedPledgedPlaces.countries].filter(c=>!!c).sort().join(', ') }}</b>
      </p>
    </div>
  </div>
</template>

<script>
import UsaMap from '@/components/element/UsaMap'
import { uniq, uniqBy, flatMap } from 'lodash'
import FamilyPledging from '@/utilities/FamilyPledging'

export default {
  components: {
    UsaMap,
  },
  props: {
    program: {
      type: Object,
      default: null,
    },
  },
  data: () => ({
    selectedPledgedPlaces: {
      states: [],
      countries: [],
    },
  }),
  computed: {
    classrooms () {
      return uniqBy(flatMap(this.program.participants, participant => {
        const classroom = participant.participant_info.classroom
        return classroom
      }), 'id').sort((a, b) => a.name > b.name ? 1 : -1)
    },
    familyPledgingObj () {
      return new FamilyPledging([this.program]
        , this.program.participants[0].id)
    },
    participant () {
      const familyPledgingObj = this.familyPledgingObj

      if (familyPledgingObj.isFamilyPledgingEnabled() && familyPledgingObj.getSmallestIdParticipantInCurrentParticipantProgram()) {
        return familyPledgingObj.getSmallestIdParticipantInCurrentParticipantProgram()
      } else if (familyPledgingObj.getCurrentParticipant()) {
        return familyPledgingObj.getCurrentParticipant()
      } else {
        return null
      }
    },
    myPledgedStates () {
      const includedStatuses = [2, 3, 8]

      return uniq(
        this.participant.participant_info.pledges
          .filter(pledge => includedStatuses.includes(pledge.pledge_status_id))
          .filter(pledge => pledge.deleted === 0)
          .map(pledge => {
            if (pledge.pledge_sponsor) {
              return pledge.pledge_sponsor.state
            }
          }),
      )
    },
    myPledgedCountries () {
      const includedStatuses = [2, 3, 8]

      return uniq(
        this.participant.participant_info.pledges
          .filter(pledge => includedStatuses.includes(pledge.pledge_status_id))
          .filter(pledge => pledge.deleted === 0)
          .filter(country => !!country)
          .flatMap(pledges => {
            if (pledges.pledge_sponsor.country_entity) {
              return pledges.pledge_sponsor.country_entity.name
            }
            return ''
          }),
      )
    },
    myPledgedPlaces () {
      return {
        states: this.myPledgedStates,
        countries: this.myPledgedCountries,
      }
    },
    statesCount () {
      const removedStates = ['DC', '']
      return this.selectedPledgedPlaces.states
        .filter((state) => state != null)
        .filter(state => !removedStates.includes(state))
        .length
    },
  },
  mounted () {
    this.selectedPledgedPlaces = this.program.pledgedPlaces
  },
  methods: {
    statesByClassroomLabel (classroom) {
      return classroom.name + (classroom.grade.name === 'Other' ? '' : ' - ' + classroom.grade.name)
    },
    trackSelect (event) {
      const label = event.target.selectedOptions[0].getAttribute('data-label')
      this.gaEvent('pledges by state map', label, 'click')
    },
  },

}
</script>
