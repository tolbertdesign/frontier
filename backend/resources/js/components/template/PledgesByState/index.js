import PledgesByState from './PledgesByState'

export default PledgesByState
