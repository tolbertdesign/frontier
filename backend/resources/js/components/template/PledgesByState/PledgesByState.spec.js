import { shallowMount } from '@vue/test-utils'
import PledgesByState from './PledgesByState.vue'

describe('PledgesByState', () => {
  const programWithOnePledge = {
    program_pledge_setting: {
      family_pledging_enabled: 1,
    },
    participants: [
      {
        participant_info: {
          family_pledging_enabled: 1,
          pledges: [
            {
              id: 1,
              pledge_status_id: 2,
              pledge_sponsor: {
                country: 'UK',
                country_entity: {
                  name: 'United Kingdom',
                },
                state: 'GA',
              },
              deleted: 0,
            },
          ],
          classroom: {
            name: 'Great Classroom Name',
            grade: {
              name: 'first grade!!!',
            },
          },
        },
      },
    ],
    pledgedPlaces: {
      states: ['GA'],
      countries: ['United Kingdom'],
    },
  }

  const programWithManyPledges = {
    program_pledge_setting: {
      family_pledging_enabled: 1,
    },
    participants: [
      {
        participant_info: {
          family_pledging_enabled: 1,
          pledges: [
            {
              id: 1,
              pledge_status_id: 2,
              pledge_sponsor: {
                country: 'US',
                country_entity: {
                  name: 'United States',
                },
                state: 'NY',
              },
              deleted: 0,
            },
            {
              id: 2,
              pledge_status_id: 2,
              pledge_sponsor: {
                country: 'UK',
                country_entity: {
                  name: 'United Kingdom',
                },
                state: 'GA',
              },
              deleted: 0,
            },
          ],
          classroom: {
            name: 'Great Classroom Name',
            grade: {
              name: 'first grade!!!',
            },
          },
        },
      },
    ],
    pledgedPlaces: {
      states: ['GA', 'AL'],
      countries: ['United States', 'United Kingdom'],
    },
  }

  const programWithManyClassrooms = {
    program_pledge_setting: {
      family_pledging_enabled: 1,
    },
    participants: [
      {
        participant_info: {
          family_pledging_enabled: 1,
          pledges: [
            {
              id: 1,
              pledge_status_id: 2,
              pledge_sponsor: {
                country: 'UK',
                country_entity: {
                  name: 'United Kingdom',
                },
                state: 'GA',
              },
              deleted: 0,
            },
          ],
          classroom: {
            id: 1,
            name: 'First Classroom Name',
            grade: {
              name: 'first grade!!!',
            },
          },
        },
      },
      {
        participant_info: {
          family_pledging_enabled: 1,
          pledges: [
            {
              id: 2,
              pledge_status_id: 2,
              pledge_sponsor: {
                country: 'UK',
                country_entity: {
                  name: 'United Kingdom',
                },
                state: 'GA',
              },
              deleted: 0,
            },
          ],
          classroom: {
            id: 2,
            name: 'Second Classroom Name',
            grade: {
              name: 'Second grade!!!',
            },
          },
        },
      },
    ],
    pledgedPlaces: {
      states: ['GA', 'AL'],
      countries: ['United States', 'United Kingdom'],
    },
  }
  const $store = {
    state: {
      lang: {
        pledge_info: {},
      },
    },
  }

  it('matches snapshot with 1 pledge', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithOnePledge },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('matches snapshot with multiple pledges', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithManyPledges },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('counts 1 state', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithOnePledge },
    })
    expect(wrapper.vm.statesCount).toBe(1)
  })

  it('counts 2 states', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithManyPledges },
    })
    expect(wrapper.vm.statesCount).toBe(2)
  })

  it('finds 1 country', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithOnePledge },
    })
    expect(wrapper.vm.myPledgedCountries)
      .toStrictEqual(['United Kingdom'])
  })

  it('finds 2 countries', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithManyPledges },
    })
    expect(wrapper.vm.myPledgedCountries)
      .toStrictEqual(['United States', 'United Kingdom'])
  })

  it('makes a classroom label', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithManyPledges },
    })
    const classroom = {
      name: 'Classname',
      grade: {
        name: 'Gradename',
      },
    }
    expect(wrapper.vm.statesByClassroomLabel(classroom))
      .toBe('Classname - Gradename')
  })

  it('finds 1 classroom', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithManyPledges },
    })
    const classroom = programWithManyPledges.participants[0].participant_info.classroom
    expect(wrapper.vm.classrooms).toStrictEqual([classroom])
  })

  it('finds 2 classrooms', () => {
    const wrapper = shallowMount(PledgesByState, {
      mocks: { $store },
      propsData: { program: programWithManyClassrooms },
    })
    const classroom1 = programWithManyClassrooms.participants[0].participant_info.classroom
    const classroom2 = programWithManyClassrooms.participants[1].participant_info.classroom
    expect(wrapper.vm.classrooms).toStrictEqual([classroom1, classroom2])
  })
})
