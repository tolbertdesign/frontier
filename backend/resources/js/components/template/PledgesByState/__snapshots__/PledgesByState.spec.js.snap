// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PledgesByState matches snapshot with 1 pledge 1`] = `
<div>
  <div class="mb-1 py-8 rounded-lg ">
    <usamap-stub activestates="" class="text-grey-light"></usamap-stub>
  </div>
  <div class="flex flex-col items-center"><span class="text-2xl"><span class="font-light">States:</span> <span class="font-bold">0</span></span>
    <div class="relative"><select class="block appearance-none w-full py-1 px-4 pr-8 border border-primary bg-white rounded-md leading-tight focus:outline-none focus:bg-white focus:border-green-500">
        <option data-label="My States" value="[object Object]">
          My States
        </option>
        <option data-label="Specific Classroom" value="">
          Great Classroom Name - first grade!!!
        </option>
        <option data-label="School States" value="[object Object]">
          School States
        </option>
      </select>
      <div class="pointer-events-none absolute flex items-center px-2 pt-2" style="right: 0px; top: 0px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" class="fill-current h-4 w-4">
          <path d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"></path>
        </svg></div>
    </div>
    <p class="text-sm">
      Countries: <b></b></p>
  </div>
</div>
`;

exports[`PledgesByState matches snapshot with multiple pledges 1`] = `
<div>
  <div class="mb-1 py-8 rounded-lg ">
    <usamap-stub activestates="GA,AL" class="text-grey-light"></usamap-stub>
  </div>
  <div class="flex flex-col items-center"><span class="text-2xl"><span class="font-light">States:</span> <span class="font-bold">2</span></span>
    <div class="relative"><select class="block appearance-none w-full py-1 px-4 pr-8 border border-primary bg-white rounded-md leading-tight focus:outline-none focus:bg-white focus:border-green-500">
        <option data-label="My States" value="[object Object]">
          My States
        </option>
        <option data-label="Specific Classroom" value="">
          Great Classroom Name - first grade!!!
        </option>
        <option data-label="School States" value="[object Object]">
          School States
        </option>
      </select>
      <div class="pointer-events-none absolute flex items-center px-2 pt-2" style="right: 0px; top: 0px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" class="fill-current h-4 w-4">
          <path d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"></path>
        </svg></div>
    </div>
    <p class="text-sm">
      Countries: <b>United Kingdom, United States</b></p>
  </div>
</div>
`;
